neptune-core/src/models/state/mod.rs

```diff
@@ -873,6 +874,54 @@ impl GlobalState {
         }
     }
 
+    /// Returns true iff the block proposal exceeds the previous proposal with
+    /// enough guesser reward to warrant a restart of the guesser with the
+    /// overhead that that brings.
+    ///
+    /// Caller must have validated the new block proposal before calling this
+    /// function.
+    ///
+    /// # Panics
+    ///
+    /// - If the new block proposal has a negative fee.
+    pub(crate) fn block_proposal_warrants_guess_restart(&self, new_proposal: &Block) -> bool {
+        match self.mining_state.mining_status {
+            // If a proposal is already being worked on, check if a
+            // re-calculation of the preprocess-data is worth it.
+            MiningStatus::Guessing(guessing_work_info) => {
+                let old_guesser_reward = guessing_work_info.total_guesser_fee;
+                let new_guesser_reward = new_proposal
+                    .total_guesser_reward()
+                    .expect("New proposal must be valid");
+                let Some(delta) = new_guesser_reward.checked_sub(&old_guesser_reward) else {
+                    return false;
+                };
+
+                let relative_delta = delta.to_nau_f64() / old_guesser_reward.to_nau_f64();
+                relative_delta >= self.cli.minimum_guesser_improvement_fraction
+            }
+
+            // If currently composing, don't interrupt this process
+            MiningStatus::Composing(_) => false,
+
+            // If no proposal is being worked on, get started on one.
+            MiningStatus::Inactive => true,
+        }
+    }
+
+    /// Returns true iff the current block proposal is present and has a high
+    /// enough guesser fee to be worth guessing on. Always  returns true if
+    /// proposal was made locally.
+    pub(crate) fn current_block_proposal_meets_threshold(&self) -> bool {
```

Comment:
It is a bad idea to have a function whose name is a sentence describing the predicate it computes, compute a different predicate. Specifically, this function returns true if the current block proposal is an own composition that does not meet the threshold, whereas the function name suggests that it should return false. That's a recipe for mistakes.